openshift-auth-proxy.js

```diff
@@ -17,89 +17,158 @@ var argv = require('yargs')
     }, 'session-ephemeral': {
       type: 'boolean',
       describe: 'Delete cookies on browser close',
-      demand: true,
-      default: false
+      default: true
     }, 'callback-url': {
       describe: 'oAuth callback URL',
-      demand: true,
-      default: '/auth/openshift/callback'
+      default: process.env.OAP_CALLBACK_URL || '/auth/openshift/callback'
     }, 'client-id': {
       describe: 'OAuth client ID',
-      demand: true
+      default: process.env.OAP_CLIENT_ID
     }, 'client-secret': {
-      describe: 'OAuth client secret',
-      demand: true
-    }, 'openshift-master': {
-      describe: 'OpenShift master to authenticate against',
-      demand: true,
-      default: 'https://kubernetes.default.svc'
-    }, 'openshift-public-master': {
-      describe: 'Public master address to redirect to',
-      demand: true
-    }, 'openshift-ca': {
-      describe: 'CA certificate[s] to use',
-      demand: true
-    }, 'tls-cert': {
+      describe: 'File containing OAuth client secret',
+      default: process.env.OAP_CLIENT_SECRET_FILE || 'secret/client-secret'
+    }, 'master-url': {
+      describe: 'Internal master address proxy will authenticate against',
+      default: process.env.OAP_MASTER_URL || 'https://kubernetes.default.svc.cluster.local:8443'
+    }, 'public-master-url': {
+      describe: 'Public master address for redirecting clients to',
+      default: process.env.OAP_PUBLIC_MASTER_URL
+    }, 'master-ca': {
+      describe: 'CA certificate(s) file to validate connection to the master',
+      default: process.env.OAP_MASTER_CA_FILE || 'secret/master-ca'
+    }, 'proxy-cert': {
       describe: 'Certificate file to use to listen for TLS',
-      demand: true
-    }, 'tls-key': {
+      default: process.env.OAP_PROXY_CERT_FILE || 'secret/proxy-cert'
+    }, 'proxy-key': {
       describe: 'Key file to use to listen for TLS',
-      demand: true
+      default: process.env.OAP_PROXY_KEY_FILE || 'secret/proxy-key'
+    }, 'proxy-tlsopts-file': {
+      describe: 'File containing JSON for proxy TLS options',
+      default: process.env.OAP_PROXY_TLS_FILE || 'secret/proxy-tls.json'
+    }, 'mutual-tls-ca': {
+      describe: 'CA cert file to use for validating TLS client certs under "mutual_tls" auth method',
+      default: process.env.OAP_PROXY_CA_FILE || 'secret/proxy-ca'
+    }, debug: {
+      describe: 'Show extra debug information at startup and during operations',
+      type: 'boolean',
+      default: process.env.OAP_DEBUG
     }
   })
+  .check(function(args) {
+    // yarg#demand doesn't work if we also provide a default, so check for missing args here
+    ["backend", 'proxy-cert', 'proxy-key', 'proxy-tlsopts-file', "session-secret",
+     "client-id", "client-secret", 'master-url', 'public-master-url', "master-ca",
+     "mutual-tls-ca", 'callback-url'].forEach( function(val) {
+      if(args[val] == null || args[val].length == 0) {
+        throw("No value specified for parameter " + val);
```

Comment:
Could you validate all parameters & throw an error with all missing or invalid parameters? Otherwise a bit annoying adding one then finding another error, etc.